### PR TITLE
Add url /xcat to retrieve nodes from xCAT db

### PR DESCRIPTION
**Pull request by zet809**
This PR add an URL /xcat for confluent to retrieve nodes from xCAT database

To setup confluent server and confluent client (Test with RH7.5-alt ppc64le with python2.7.5)

Prepare venv:
```
yum install libffi-devel.ppc64le openssl-devel libxml2-devel libxslt-devel
#  For postgrepsql
yum install postgresql-devel
# For python2.7, install virtualenv
pip install virtualenv
virtualenv /opt/confluent/venv
source /opt/confluent/venv/bin/activate
pip install crypto pyasn1 pycrypto eventlet pyparsing netifaces scrapy pysnmp paramiko pyghmi pyte pycryptodome sqlalchemy psycopg2
deactivate
```
Build and install confluent server and confluent client
```
git clone https://github.com/xcat2/confluent.git
cd confluent/; git checkout add_xcat_interfaces; cd -
cd confluent/confluent_server/; ./buildrpm; cd -
cd confluent/confluent_client/; ./buildrpm; cd -
rpm -ivh /root/rpmbuild/RPMS/noarch/confluent_server-2.1.0.* --nodeps
rpm -ivh /root/rpmbuild/RPMS/noarch/confluent_client-2.1.0.* --nodeps
```
Start confluent:
```
service confluent start
```
To connect to remote pgsql DB, need to provide cfglog like this:
```
# cat /etc/xcat/cfgloc
Pg:dbname=xcatdb;host=<db_host>|<db_user>|<db_pass>
```
And on db host, for Postgresql, add the line below in /var/lib/pgsql/data/pg_hba.conf, and `service postgresql restart`
```
host    all          all        <client_ip>/32      md5
```

The output below shows connecting to our briggs01 MN, `TEST_XCAT` is need to access the /xcat url
```
# TEST_XCAT=yes /opt/confluent/bin/nodelist
abcdef
bmc=|\D+(\d+)\D+(\d+)\D+(\d+)|172.12.(($1-1)*32+$2-1).($3%100+100)|
bogusgroup
c461c022
c461c026
cn1
cn_regex
compute
coral_sn_regex
coral_sn_regex1
core01
edge_switch
frame5u20
lpar5
mgmtsw01
mid05
mid05to12cn13
mid05tor12
mid05tor12cn01
mid05tor12cn02
mid05tor12cn02k02
mid05tor12cn03
mid05tor12cn04
mid05tor12cn05
mid05tor12cn05-bmc
mid05tor12cn05-bmc-test
mid05tor12cn06
mid05tor12cn07
mid05tor12cn08
mid05tor12cn09
mid05tor12cn10
mid05tor12cn11
mid05tor12cn12
mid05tor12cn13
mid05tor12cn14
mid05tor12cn15
mid05tor12cn15-ipmi
mid05tor12cn16
mid05tor12cn16-ipmi
mid05tor12cn17
mid05tor12cn18
mid05tor25
mid_switch
mlnx
node0001
node0002
node-8335-gth-1318c4a
openbmc_test
service
service[1-5]
sn01
sn02
testnode
testnodes
victor
xcatdefaults
```

### CODE REVIEW

**Comment by zet809** [comment]
The confluent is enabled on 10.3.5.3, you can use http://10.3.5.3:4005/ to take a try.
```
# curl http://10.3.5.3:4005/xcat/nodes -u xcat
Enter host password for user 'xcat':
<html><head><meta charset="UTF-8"><title>Confluent REST Explorer: /xcat/nodes</title></head><body><form action="./nodes" method="post">Only fields that have their boxes checked will have their respective values honored by the confluent server.<hr><input type="hidden" name="restexplorerhonorkey" value=""><a rel="self" href="./nodes">./nodes</a><br><a rel="collection" href="./">./</a><br><a href="p8cn02/" rel="item">p8cn02/</a> . . . . . . . . . . . . <button type="submit" name="restexplorerop" value="delete" formaction="p8cn02/">delete</button><br><input value="update" name="restexplorerop" type="submit"></form></body></html>
```
Or
```
# TEST_XCAT=1 /opt/confluent/bin/nodelist
p8cn02
# /opt/confluent/bin/nodelist
node-test01
node-test02
```

**Comment by zet809** [comment]
Hi, @jjohnson42 , will you pls help to take a look at whether it is the correct logic to add new url as the PR. Thank you very much!

File: confluent_server/confluent/xcclient/dbsession.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python
+###############################################################################
+# IBM(c) 2007 EPL license http://www.eclipse.org/legal/epl-v10.html
```
Comment (jjohnson42) [inline confluent_server/confluent/xcclient/dbsession.py]: Do you mean 2007?

File: confluent_server/confluent/xcclient/utils.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python
+###############################################################################
+# IBM(c) 2007 EPL license http://www.eclipse.org/legal/epl-v10.html
+###############################################################################
```
Comment (jjohnson42) [inline confluent_server/confluent/xcclient/utils.py]: You mean 2019 I assume..

**Comment by jjohnson42** [comment]
I'm going to have to warn in advance I'm tied up until tuesday, so sorry that there will be some delay in a more thorough examination.

**Comment by jjohnson42** [comment]
FYI:
`# curl http://10.3.5.3:4005/xcat/nodes -u xcat`

You could do either:
`# curl http://10.3.5.3:4005/xcat/nodes.json -u xcat`
Or:
-H 'Accept: application/json' http://10.3.5.3:4005/xcat/nodes

To get json from http api.

**Comment by jjohnson42** [comment]
BTW, is this just an experiment to see what adding urls would be like or part of a plan to add on functionality?  Didn't want to merge if it's just testing the waters.

I'm inclined to the root url being '/integrations/xcat2' or similar, in anticipation of having some more integrations.

Also, potentially relevant, here is the confluent native service node operation:
https://hpc.lenovo.com/users/documentation/collective.html

File: confluent_server/confluent/xcclient/attributes.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    'hardwaremanagement.method': 'nodehm.mgt',
+    'enclosure.bay': None,
+    'enclosure.extends': None,
+    'enclosure.manager': None,
+    'id.model': 'vpd.mtm',
+    'id.serial': 'vpd.serial',
+    'id.uuid': 'vpd.uuid',
+    'net.bootable': None, 
+    'net.ipv4_gateway': None,
+    'net.hwaddr': 'mac.mac',
```
Comment (jjohnson42) [inline confluent_server/confluent/xcclient/attributes.py]: Though you may not want to model this, it is a good time to highlight that 'net' attributes are special, as they can be sub-instanced to model multiple interfaces:
# nodeattrib d3 net*switch
d3: net.compute.switch: computesw1
d3: net.mgt.switch: mgtsw1
d3: net.switch:

Contrast with the mac model where multiple macs are crammed into a single field.

File: confluent_server/confluent/xcclient/attributes.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    'groups': 'nodelist.groups',
+    'collective.manager': 'nodehm.mgt',
+    'discovery.passwordrules': None,
+    'discovery.policy': ['switch', 'mtms', 'sequential'],
+    'info.note': 'nodelist.comments',
+    'location.room': 'nodepos.room',
+    'location.row': 'nodepos.comments', 
+    'location.rack': 'nodepos.rack',
+    'location.u': 'nodepos.u',
+    'console.logging': 'site.consoleondemand',
```
Comment (jjohnson42) [inline confluent_server/confluent/xcclient/attributes.py]: While the intent is the same, the values are different, e.g. the console.logging values are 'full', 'interactive', and 'none'.

**Comment by jjohnson42** [comment]
I'll again ask what the intent is.  For reference we have much better performance and the service node model is a bit easier.  Is it for integration for code to reach back to an xcat db rather than doing makeconfluentcfg as a transient?  In that case I'm not certain the intent of trying to map the attribute values.

**Comment by zet809** [comment]
Thank you very much for your comments, @jjohnson42 , will reply later tomorrow.

File: confluent_server/confluent/xcclient/utils.py (hunk 2 above)
Comment (zet809) [reply confluent_server/confluent/xcclient/utils.py]: updated

File: confluent_server/confluent/xcclient/dbsession.py (hunk 1 above)
Comment (zet809) [reply confluent_server/confluent/xcclient/dbsession.py]: updated

**Pull request closed without merging**